spanner/metrics.go

```diff
@@ -0,0 +1,474 @@
+	attrKeyValues = append(attrKeyValues, attribute.String(metricLabelKeyDirectPathUsed, strconv.FormatBool(mt.currOp.currAttempt.directPathUsed)))
+
+	rpcStatus := mt.currOp.status
+	if mDetails.recordedPerAttempt {
+		rpcStatus = mt.currOp.currAttempt.status
+	}
+
+	// Add additional attributes to metrics
+	for _, attrKey := range mDetails.additionalAttrs {
+		switch attrKey {
+		case metricLabelKeyStatus:
+			attrKeyValues = append(attrKeyValues, attribute.String(metricLabelKeyStatus, rpcStatus))
+		default:
+			return attrKeyValues, fmt.Errorf("unknown additional attribute: %v", attrKey)
+		}
+	}
+
+	return attrKeyValues, nil
+}
+
+// Convert error to grpc status error
+func convertToGrpcStatusErr(err error) (codes.Code, error) {
+	if err == nil {
+		return codes.OK, nil
+	}
+
+	if errStatus, ok := status.FromError(err); ok {
+		return errStatus.Code(), status.Error(errStatus.Code(), errStatus.Message())
+	}
+
+	ctxStatus := status.FromContextError(err)
+	if ctxStatus.Code() != codes.Unknown {
+		return ctxStatus.Code(), status.Error(ctxStatus.Code(), ctxStatus.Message())
+	}
+
+	return codes.Unknown, err
+}
+
+// recordAttemptCompletion records as many attempt specific metrics as it can
+// Ignore errors seen while creating metric attributes since metric can still
+// be recorded with rest of the attributes
+func recordAttemptCompletion(mt *builtinMetricsTracer) {
+	if !mt.builtInEnabled {
+		return
+	}
+
+	// Calculate elapsed time
+	elapsedTime := convertToMs(time.Since(mt.currOp.currAttempt.startTime))
+
+	// Record attempt_latencies
+	attemptLatAttrs, _ := mt.toOtelMetricAttrs(metricNameAttemptLatencies)
+	mt.instrumentAttemptLatencies.Record(mt.ctx, elapsedTime, metric.WithAttributes(attemptLatAttrs...))
+}
+
+// recordOperationCompletion records as many operation specific metrics as it can
+// Ignores error seen while creating metric attributes since metric can still
+// be recorded with rest of the attributes
+func recordOperationCompletion(mt *builtinMetricsTracer) {
+	if !mt.builtInEnabled {
+		return
+	}
+
+	// Calculate elapsed time
+	elapsedTimeMs := convertToMs(time.Since(mt.currOp.startTime))
+
+	// Record operation_count
+	opCntAttrs, _ := mt.toOtelMetricAttrs(metricNameOperationCount)
+	mt.instrumentOperationCount.Add(mt.ctx, 1, metric.WithAttributes(opCntAttrs...))
+
+	// Record operation_latencies
+	opLatAttrs, _ := mt.toOtelMetricAttrs(metricNameOperationLatencies)
+	mt.instrumentOperationLatencies.Record(mt.ctx, elapsedTimeMs, metric.WithAttributes(opLatAttrs...))
+
+	// Record attempt_count
+	attemptCntAttrs, _ := mt.toOtelMetricAttrs(metricNameAttemptCount)
+	if mt.currOp.attemptCount > 1 {
+		// Only record when retry count is greater than 0 so the retry
+		// graph will be less confusing
+		mt.instrumentAttemptCount.Add(mt.ctx, mt.currOp.attemptCount-1, metric.WithAttributes(attemptCntAttrs...))
+	}
+}
+func convertToMs(d time.Duration) float64 {
+	return float64(d.Nanoseconds()) / 1000000
```

Comment:
Should we use `return float64(d.Nanoseconds()) / float64(time.Millisecond)`?

Reply:
Done